plugins/sentinelone/komand_sentinelone/tasks/monitor_activities_and_events/task.py

```diff
@@ -0,0 +1,327 @@
+            cursors = self.get_cursors(state)
+            total_queries = self.count_queries(queries)
+            total_forbidden_responses = 0
+            last_run_timestamp = state.get(LAST_RUN_TIMESTAMP)
+            self.logger.info(cursors)
+            is_paginating = any(cursors.values())
+            if is_paginating:
+                self.log_pagination_cycle(cursors)
+
+            current_run_timestamp, lookback_timestamp = self.get_lookback_values(
+                is_initial_run, custom_config, last_run_timestamp
+            )
+
+            all_logs, total_forbidden_responses = self.collect_logs_handler(
+                queries,
+                cursors,
+                state,
+                lookback_timestamp,
+                current_run_timestamp,
+                total_forbidden_responses,
+                is_paginating,
+            )
+
+            # Check if all ran queries have returned 401 or 403 errors and raise an exception if so
+            if total_forbidden_responses >= total_queries > 0:
+                return [], existing_state, False, 401, PluginException(preset=PluginException.Preset.UNAUTHORIZED)
+            has_more_pages = self.determine_next_pagination_cycle(state, lookback_timestamp, current_run_timestamp)
+            return all_logs, state, has_more_pages, 200, None
+        except ApiException as error:
+            return [], existing_state, False, error.status_code, error
+        except Exception as error:
+            return [], existing_state, False, 500, PluginException(preset=PluginException.Preset.UNKNOWN, data=error)
+
+    def check_initial_run(self, state: Dict) -> bool:
+        """
+        Check if state exists, return False if so indicating initial run
+        """
+        is_initial_run = not state
+        if is_initial_run:
+            self.logger.info("No state detected. Instantiating initial run.")
+        else:
+            self.logger.info("State detected. Instantiating continuation run.")
+        return is_initial_run
+
+    def check_queries(self, params: Dict) -> Tuple[str, str, str]:
+        """
+        Return whether to query a specific endpoint based on the input parameters
+        """
+        return {
+            ACTIVITIES_LOGS: params.get(Input.COLLECTACTIVITIES),
+            EVENTS_LOGS: params.get(Input.COLLECTEVENTS),
+            THREATS_LOGS: params.get(Input.COLLECTTHREATS),
+        }
+
+    def get_cursors(self, state: Dict) -> Tuple[str, str, str]:
+        """
+        Return existing cursors for each endpoint
+        """
+        return {
+            ACTIVITIES_LOGS: state.get(ACTIVITIES_PAGE_CURSOR),
+            EVENTS_LOGS: state.get(EVENTS_PAGE_CURSOR),
+            THREATS_LOGS: state.get(THREATS_PAGE_CURSOR),
+        }
+
+    def count_queries(self, queries: dict) -> int:
+        "Count how many queries are to be made and return the value"
+        return len(list(filter(lambda query: query is not None, queries.values())))
```

Comment:
Maybe worth logging out the number of queries to be run?

Reply:
I've combined this and the below into one log.